# Market Basket
<p/>
auther: Michael Stephenson


### Part I: Research Question

#### A1. Proposal of Question
<p/>
The question this analysis preposes is whether market basket analysis would be suitable for gaining helpful information about store products and customer behavior from a dataset of products?

#### A2. Defined Goal
<p/>
The goal of this case study will be to find all the useful metrics concerning the provided dataset using Rules Based Analysis. It is necessary to understand the practical signifinace of using rules gain insignt about itemsets the algorithm finds. 


### Part II: Market Basket Justification


#### B1. Explanation of Market Basket

<p/>

Market Basket Analysis is one of the essential techniques used by large retailers to uncover associations between items. It works by looking for combinations of things that frequently occur in transactions. Association Rules are widely used to analyze retail basket or transaction data. Interestingness measures identify strong rules discovered in transaction data. Support indicates how frequently the item set appears in the data set. Expected confidence is the probability of occurrence of the consequent given the antecedent. The Lift tells us how much better a rule is at predicting the result. (Shah 2022)

<p/>

Market basket analysis is a machine learning tool. Association Rules are also fundamental concepts of Market basket analysis using machine learning. Rules do not extract an individual's preference; they find relationships between a set of elements. Relationships are what makes them different from collaborative filtering. Collaborative filtering reverts all transactions associated with a user ID to identify likenesses between user preferences. User ID helps recommend items on e-commerce websites, songs on Spotify, etc. Rules do not collect user transactions to determine relationships. The support measure gives the frequency an itemset is in transactions. Many transactions will have Bread on the cart with Butter, but Bread with Shampoo? Maybe not so many. Support is the fraction of total transactions the itemset occurs in. The value of support will have, the identifed rules to consider for analysis; a rule to consider only the itemsets that occur at least 50 times out of 10,000 transactions, for example. If an itemset happens to have shallow support, the Market Basket cannot make an association; there is not enough information about the relationship between its items. The confidence measures the likeliness of future item results on a cart, given that the cart already has earlier transacted things. (Garg 2018)

<p/>

We can say by the transacted knowledge that a high confidence rule will exist in the itemset; Captain Crunch and Milk. Since confidence is the conditional probability of occurrence, confidence for an association rule with a persistent consequence will always be high. Let us consider the confidence for Butter and Bread? What is the conditional probability Butter will also have Bread? Very high because the item set is a persistent consequence. Lift measures the frequency of an itemset; while calculating the confidence of Y given X., Supermarkets will have thousands of different products in-store; so finding lift values for each item will get computationally expensive. How do you come up with a set of most important association rules to be considered? The Apriori algorithm. (Garg 2018)

<p/>

Since transactional datasets can hypothetically be enormous, both in terms of the number of transactions and in the number of items purchased, computational cost interrogating large datasets is high, but the Apriori algorithm generates association rules from the transactional data because it reduces the number of itemsets to evaluate. The algorithm utilizes a prior belief about the properties of frequent itemsets called the Apriori Property, which states that all subsets of frequent itemsets must also be frequent. (Junaid 2019)



#### B2. Transaction Example

<p/>

We can look at transaction data from the teleco_market_basket dataset to demonstrate a typical market basket itemset below. The items make it easy to infer the customer is purchasing peripherals for an Apple laptop.

<p/>

Example Transaction:

<p/>

Apple-Pencil SanDisk-Ultra-128GB-card RUNMUS-Gaming-Headset TopMate-C5-Laptop-Cooler-pad

<p/>



#### B3. Market Basket Assumption

<p/>

Note: An assumption of rules based association is that it can identify patterns in customer behavior. (Market basket Analysis 2021)



### Part III: Data Preparation and Analysis



#### C1. Transforming the Dataset

<p/>

The data is made up of continuous and categorical data. Missing data will only be dropped. NAN features will otherwise be converted to zero.
<p/>

#### C-1a. Initial data classification: 

<p/>
    
Categorical Predictor Data:
    
<p/>
    
<ul>
<li>Education</li>
<li>Employment</li>
<li>Marital</li>
<li>Gender</li>
<li>Churn</li>
<li>Techie</li>
<li>Contract</li>
<li>Port_modem</li>
<li>Tablet</li>
<li>InternetService</li>
<li>Phone</li>
<li>Multiple</li>
<li>OnlineSecurity</li>
<li>OnlineBackup</li>
<li>DeviceProtection</li>
<li>TechSupport</li>
<li>StreamingTV</li>
<li>StreamingMovies</li>
<li>PaperlessBilling</li>
<li>PaymentMethod</li>
<li>Multiple</li>
</ul>
    
<p/>

<p/>
    
Continuous Predictor Data <p/>
<ul>
<li> Children</li>
<li> Age</li>
<li> Income</li>
<li> Outage_sec_perweek</li>
<li> Tenure</li>
<li> MonthlyCharge</li>
<li> Bandwidth_GB_Year</li>
<li>Multiple</li>
</ul>
    
<p/>

<p/>
    
Discrete Ordinal Predictor Data 
    
<p/>
<ul>
<li>Yearly_equip_failure</li>
<li>Timely_Response</li>
<li>Timely_Fixes</li>
<li>Timely_Replacements</li>
<li>Respectful_Response</li>
<li>Options</li>
<li> Courteous_Exchange</li>
<li>Active_Listening</li>
</ul>
    
<p/>


#### C-1b. Steps to prepare the data for the analysis:


In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline 
# Ignore Warning Code
import warnings
warnings.filterwarnings('ignore')

In [2]:
# import data set
dataset = pd.read_csv('teleco_market_basket.csv')
dataset

Item01  \
0                                        NaN   
1               Logitech M510 Wireless mouse   
2                                        NaN   
3      Apple Lightning to Digital AV Adapter   
4                                        NaN   
...                                      ...   
14997         Falcon Dust Off Compressed Gas   
14998                                    NaN   
14999                            HP 63XL Ink   
15000                                    NaN   
15001                           Apple Pencil   

                                 Item02                 Item03  \
0                                   NaN                    NaN   
1                             HP 63 Ink              HP 65 ink   
2                                   NaN                    NaN   
3      TP-Link AC1750 Smart WiFi Router           Apple Pencil   
4                                   NaN                    NaN   
...                                 ...                    ...   
14997                               NaN                    NaN   
14998                               NaN                    NaN   
14999         Apple USB-C Charger cable                    NaN   
15000                               NaN                    NaN   
15001          SanDisk Ultra 128GB card  RUNMUS Gaming Headset   

                             Item04                     Item05        Item06  \
0                               NaN                        NaN           NaN   
1        nonda USB C to USB Adapter  10ft iPHone Charger Cable  HP 902XL ink   
2                               NaN                        NaN           NaN   
3                               NaN                        NaN           NaN   
4                               NaN                        NaN           NaN   
...                             ...                        ...           ...   
14997                           NaN                        NaN           NaN   
14998                           NaN                        NaN           NaN   
14999                           NaN                        NaN           NaN   
15000                           NaN                        NaN           NaN   
15001  TopMate C5 Laptop Cooler pad                        NaN           NaN   

                             Item07                               Item08  \
0                               NaN                                  NaN   
1      Creative Pebble 2.0 Speakers  Cleaning Gel Universal Dust Cleaner   
2                               NaN                                  NaN   
3                               NaN                                  NaN   
4                               NaN                                  NaN   
...                             ...                                  ...   
14997                           NaN                                  NaN   
14998                           NaN                                  NaN   
14999                           NaN                                  NaN   
15000                           NaN                                  NaN   
15001                           NaN                                  NaN   

                              Item09                                   Item10  \
0                                NaN                                      NaN   
1      Micro Center 32GB Memory card  YUNSONG 3pack 6ft Nylon Lightning Cable   
2                                NaN                                      NaN   
3                                NaN                                      NaN   
4                                NaN                                      NaN   
...                              ...                                      ...   
14997                            NaN                                      NaN   
14998                            NaN                                      NaN   
14999                            NaN                             

In [3]:
# Describe dataset 
dataset.describe()

Item01                          Item02  \
count                             7501                            5747   
unique                             115                             117   
top     Dust-Off Compressed Gas 2 pack  Dust-Off Compressed Gas 2 pack   
freq                               577                             484   

                                Item03                          Item04  \
count                             4389                            3345   
unique                             115                             114   
top     Dust-Off Compressed Gas 2 pack  Dust-Off Compressed Gas 2 pack   
freq                               375                             201   

                           Item05                 Item06  \
count                        2529                   1864   
unique                        110                    106   
top     Apple USB-C Charger cable  USB 2.0 Printer cable   
freq                          153                    107   

                           Item07                     Item08  \
count                        1369                        981   
unique                        102                         97   
top     Apple USB-C Charger cable  Apple USB-C Charger cable   
freq                           96                         67   

                           Item09                     Item10  \
count                         654                        395   
unique                         88                         80   
top     Apple USB-C Charger cable  Apple USB-C Charger cable   
freq                           57                         31   

                              Item11                     Item12  \
count                            256                        154   
unique                            66                         50   
top     TopMate C5 Laptop Cooler pad  Apple USB-C Charger cable   
freq                              22                         15   

                           Item13                     Item14  \
count                          87                         47   
unique                         43                         28   
top     Apple USB-C Charger cable  Apple USB-C Charger cable   
freq                            8                          4   

                                    Item15                          Item16  \
count                                   25                               8   
unique                                  19                               8   
top     ARRIS SURFboard SB8200 Cable Modem  3A USB Type C Cable 3 pack 6FT   
freq                                     3                               1   

                          Item17                                      Item18  \
count                          4                                           4   
unique                         3                                           3   
top     SanDisk Ultra 128GB card  Brother Genuine High Yield Toner Cartridge   
freq                           2                                           2   

                               Item19                               Item20  
count                               3                                    1  
unique                              3                                    1  
top     NETGEAR Nighthawk WiFi Router  FEIYOLD Blue light Blocking Glasses  
freq                                1                                    1

Step 1. Check null values

In [4]:
# check null values
print(dataset.isnull().sum())

Item01     7501
Item02     9255
Item03    10613
Item04    11657
Item05    12473
Item06    13138
Item07    13633
Item08    14021
Item09    14348
Item10    14607
Item11    14746
Item12    14848
Item13    14915
Item14    14955
Item15    14977
Item16    14994
Item17    14998
Item18    14998
Item19    14999
Item20    15001
dtype: int64


Step 4.  Drop all rows with Null value in all rows only

In [5]:
# Drop Null values in all rows
dataset.dropna(how='all', inplace=True)
# Review changes
dataset.head()

Item01  \
1               Logitech M510 Wireless mouse   
3      Apple Lightning to Digital AV Adapter   
5  UNEN Mfi Certified 5-pack Lightning Cable   
7                        Cat8 Ethernet Cable   
9             Dust-Off Compressed Gas 2 pack   

                             Item02                      Item03  \
1                         HP 63 Ink                   HP 65 ink   
3  TP-Link AC1750 Smart WiFi Router                Apple Pencil   
5                               NaN                         NaN   
7                         HP 65 ink                         NaN   
9     Screen Mom Screen Cleaner kit  Moread HDMI to VGA Adapter   

                       Item04                     Item05        Item06  \
1  nonda USB C to USB Adapter  10ft iPHone Charger Cable  HP 902XL ink   
3                         NaN                        NaN           NaN   
5                         NaN                        NaN           NaN   
7                         NaN                        NaN           NaN   
9       HP 62XL Tri-Color ink  Apple USB-C Charger cable           NaN   

                         Item07                               Item08  \
1  Creative Pebble 2.0 Speakers  Cleaning Gel Universal Dust Cleaner   
3                           NaN                                  NaN   
5                           NaN                                  NaN   
7                           NaN                                  NaN   
9                           NaN                                  NaN   

                          Item09                                   Item10  \
1  Micro Center 32GB Memory card  YUNSONG 3pack 6ft Nylon Lightning Cable   
3                            NaN                                      NaN   
5                            NaN                                      NaN   
7                            NaN                                      NaN   
9                            NaN                                      NaN   

                         Item11                     Item12  \
1  TopMate C5 Laptop Cooler pad  Apple USB-C Charger cable   
3                           NaN                        NaN   
5                           NaN                        NaN   
7                           NaN                        NaN   
9                           NaN                        NaN   

                         Item13                       Item14  \
1  HyperX Cloud Stinger Headset  TONOR USB Gaming Microphone   
3                           NaN                          NaN   
5                           NaN                          NaN   
7                           NaN                          NaN   
9                           NaN                          NaN   

                           Item15                          Item16  \
1  Dust-Off Compressed Gas 2 pack  3A USB Type C Cable 3 pack 6FT   
3                             NaN                             NaN   
5                             NaN                             NaN   
7                             NaN                             NaN   
9                             NaN                             NaN   

                  Item17                    Item18  \
1  HOVAMP iPhone charger  SanDisk Ultra 128GB card   
3                    NaN                       NaN   
5                    NaN                       NaN   
7                    NaN                       NaN   
9                    NaN                       NaN   

                                 Item19                               Item20  
1  FEEL2NICE 5 pack 10ft Lighning cable  FEIYOLD Blue light Blocking Glasses  
3                                   NaN                                  NaN  
5                                   NaN                                  NaN  
7                                   NaN                                  NaN  
9                                   NaN                                  NaN

Step.5 Verify no null values remain after replacing nan with 0.

In [6]:
# Converting null to 0
dataset= dataset.fillna(0)
# check null values
print(dataset.isnull().sum())

Item01    0
Item02    0
Item03    0
Item04    0
Item05    0
Item06    0
Item07    0
Item08    0
Item09    0
Item10    0
Item11    0
Item12    0
Item13    0
Item14    0
Item15    0
Item16    0
Item17    0
Item18    0
Item19    0
Item20    0
dtype: int64


Step 6. Export cleaned data

In [7]:
# export file to csv
dataset.to_csv('cleaned_teleco.csv', index = False)
dataset

Item01  \
1                   Logitech M510 Wireless mouse   
3          Apple Lightning to Digital AV Adapter   
5      UNEN Mfi Certified 5-pack Lightning Cable   
7                            Cat8 Ethernet Cable   
9                 Dust-Off Compressed Gas 2 pack   
...                                          ...   
14993               SanDisk 32GB Ultra SDHC card   
14995      Apple Lightning to Digital AV Adapter   
14997             Falcon Dust Off Compressed Gas   
14999                                HP 63XL Ink   
15001                               Apple Pencil   

                                     Item02                        Item03  \
1                                 HP 63 Ink                     HP 65 ink   
3          TP-Link AC1750 Smart WiFi Router                  Apple Pencil   
5                                         0                             0   
7                                 HP 65 ink                             0   
9             Screen Mom Screen Cleaner kit    Moread HDMI to VGA Adapter   
...                                     ...                           ...   
14993                 Vsco 70 pack stickers  SanDisk 128GB microSDXC card   
14995  Nylon Braided Lightning to USB cable                  Apple Pencil   
14997                                     0                             0   
14999             Apple USB-C Charger cable                             0   
15001              SanDisk Ultra 128GB card         RUNMUS Gaming Headset   

                             Item04                              Item05  \
1        nonda USB C to USB Adapter           10ft iPHone Charger Cable   
3                                 0                                   0   
5                                 0                                   0   
7                                 0                                   0   
9             HP 62XL Tri-Color ink           Apple USB-C Charger cable   
...                             ...                                 ...   
14993                             0                                   0   
14995         USB 2.0 Printer cable  ARRIS SURFboard SB8200 Cable Modem   
14997                             0                                   0   
14999                             0                                   0   
15001  TopMate C5 Laptop Cooler pad                                   0   

                          Item06                        Item07  \
1                   HP 902XL ink  Creative Pebble 2.0 Speakers   
3                              0                             0   
5                              0                             0   
7                              0                             0   
9                              0                             0   
...                          ...                           ...   
14993                          0                             0   
14995  Apple USB-C Charger cable                             0   
14997                          0                             0   
14999                          0                             0   
15001                          0                             0   

                                    Item08                         Item09  \
1      Cleaning Gel Universal Dust Cleaner  Micro Center 32GB Memory card   
3                                        0                              0   
5                                        0                              0   
7                                        0                              0   
9                                        0                              0   
...                                    ...                            ...   
14993                                    0                              0   
14995                                    0                              0   
14997                                    0                              0   
14999                       

#### C2. Code Execution

Step 1. Convert the dataset to list of transactions

In [8]:
# dataframe lists (Korstanje 2021)
transactions = []
for i in range(0, 7500):
    transactions.append([str(dataset.values[i,j]) for j in range(0, 20)])

Step 2. Run the apriori algorithm.

In [9]:
# run the apriori algorithm (Korstanje 2021)

from efficient_apriori import apriori 

# our min support is 10, but it has to be expressed as a percentage for efficient-apriori
min_support = 20/len(transactions) 

# min confidence allows you to delete rules with low confidence.
# For now set min_confidence = 0 to obtain all the rules
min_confidence = 0.75
itemsets, rules = apriori(transactions, min_support=min_support, min_confidence=min_confidence)


#### C3. Association Rules Table

##### Values for the support, lift, and confidence below

In [10]:
# Turning list into a new dataframe (python - Apriori rule to pandas dataframe n.d.)
data = [dict(**rule.__dict__, confidence=rule.confidence, support=rule.support, lift=rule.lift, conviction=rule.conviction)
            for rule in rules]
rules_df = pd.DataFrame(data)
rules_df.sort_values(by ='lift', ascending = False, inplace = True)
rules_df.head(10)

lhs  \
1674  (10ft iPHone Charger Cable 2 Pack, Nylon Braid...   
1671  (10ft iPHone Charger Cable 2 Pack, Nylon Braid...   
1673  (0, 10ft iPHone Charger Cable 2 Pack, Nylon Br...   
869   (SAMSUNG 128GB card, VIVO Dual LCD Monitor Des...   
742   (HyperX Cloud Stinger Headset, SanDisk Ultra 4...   
1389  (Dust-Off Compressed Gas 2 pack, HP 61 ink, St...   
586   (Falcon Dust Off Compressed Gas, Stylus Pen fo...   
371   (Apple USB-C Charger cable, Logitech MK345 Wir...   
592   (Falcon Dust Off Compressed Gas, YUNSONG 3pack...   
1180  (Apple Pencil, Apple USB-C Charger cable, Synt...   

                                      rhs  count_full  count_lhs  count_rhs  \
1674  (0, Dust-Off Compressed Gas 2 pack)          23         30       1787   
1671    (Dust-Off Compressed Gas 2 pack,)          23         30       1788   
1673    (Dust-Off Compressed Gas 2 pack,)          23         30       1788   
869                                  (0,)          34         34       7499   
742                                  (0,)          25         25       7499   
1389                                 (0,)          70         70       7499   
586                                  (0,)          68         68       7499   
371                                  (0,)          50         50       7499   
592                                  (0,)          25         25       7499   
1180                                 (0,)          25         25       7499   

      num_transactions  confidence   support      lift     conviction  
1674              7500    0.766667  0.003067  3.217683       3.264571  
1671              7500    0.766667  0.003067  3.215884       3.264000  
1673              7500    0.766667  0.003067  3.215884       3.264000  
869               7500    1.000000  0.004533  1.000133  133333.333333  
742               7500    1.000000  0.003333  1.000133  133333.333333  
1389              7500    1.000000  0.009333  1.000133  133333.333333  
586               7500    1.000000  0.009067  1.000133  133333.333333  
371               7500    1.000000  0.006667  1.000133  133333.333333  
592               7500    1.000000  0.003333  1.000133  133333.333333  
1180              7500    1.000000  0.003333  1.000133  133333.333333

In [ ]:
rules_df.shape

#### C4. Top Three Rules



##### An example the top 3 Association Rules from the above output:

<p/>

(10ft iPHone Charger Cable 2 Pack, Nylon Braid) > (0, Dust-Off Compressed Gas 2 pack) confidence 0.766667 support 0.003067 lift 3.217683	
(10ft iPHone Charger Cable 2 Pack, Nylon Braid) > (Dust-Off Compressed Gas 2 pack,) confidence 0.766667 support 0.003067 lift 3.215884	
(10ft iPHone Charger Cable 2 Pack, Nylon Braid) > (Dust-Off Compressed Gas 2 pack,) confidence 0.766667 support 0.003067 lift 3.215884


### Part IV: Data Summary and Implications



#### D1. Significance of Support, Lift, and Confidence Summary

#### D-1a. Significance of Support

Since the Apriori algorithm is foundational to basket analysis, which is the study of a shoppers basket, the goal is to find products that are often associated together or frequent itemsets; Frequent Itemset Mining. Support identifies the number of transactions an item occurs in, and our current analysis lists all possible combinations. Therefore, it is a common step to set a threshold to raise the minimum occurrence to a more statistically significant number. For this study, the support threshold is 20 with <code>min_support = 20/len(transactions)</code> in Step 2 above. (Korstanje 2021)

#### D-1b. Significance of Confidence

<p/>

Now that the frequent itemsets are known it is time to make them into Association Rules written in the format <code> product X => product Y</code>. Association Rules mean that product X directly relates to product Y in a shopping cart.  Confidence gives a percentage of correctness that items are associated together; 100% means X and Y are always together and 50% means they only occure together half the time. For this study the confidence score is set to a minimum of 75% (a statistically significant level) with <code> min_confidence = 0.75 </code>. The result is a dataset of 1739 records. (Korstanje 2021)

#### D-1c. Significance of Lift

<p/>

Lift now determines the strength of association between the rule and the products found in the basket. Lift compares the improving associations across the dataset. For example, product X  of 10% the cases that improve with a confidence of 75% improves by 7.5. Otherwise, if Lift is one, the products are independent. When the score is higher than one, Lift explains how strongly the right-side object is dependent on the left-side object. (Korstanje 2021)

<p/>

#### D2. Practical Significance of Findings

<p/>

An accuracy in market basket analysis is a decision-based approach to knowing why you prefer one metric of either support, confidence, or lift over the remaining two. All three have validity, but determining why one works over the others in a specific use case is left to the researcher. It is difficult to state that one rule has more merit than another. Conviction may help disambiguate the significance of one item in a rule versus another. A conviction of one means there is no association, like Lift, but anything more significant than one indicates a relationship and direction. Since Lift is the same for X or Y in an itemset, it has no directionality of association. A conviction can be different in X or Y, so a dominant directionality of the relationship is assumed. Practical significance will need to determine what value any of the assumption rules have in a real-world application rather than statical significance to deploy a market basket analysis in a helpful way and obtain a business solution. (Korstanje 2021) (Dobilas 2022)

#### D3. Course of Action

<p/>

We accomplished the project goal! We have established that the dataset helps identify association rules about products. The project results with itemsets representing useful metrics about customer behavior buying store products; further steps can now take place to choose an association rule with practical significance to improving company performance contained within  <code>rules_df</code>. This study is insightful enough to show how association rules can build a useful logical tool for the company to improve performance. However, a question remains about how management can best represent their needs and utilize these findings? 

### Sources:

Dobilas, S. (2022, February 5). Apriori Algorithm for Association Rule Learning — How To Find Clear Links Between Transactions. Medium. https://towardsdatascience.com/apriori-algorithm-for-association-rule-learning-how-to-find-clear-links-between-transactions-bf7ebc22cf0a

<p/>

Garg A. Complete guide to Association Rules (1/2). Towards Data Science. Published September 3, 2018. https://towardsdatascience.com/association-rules-2-aa9a77241654

<p/>

Junaid. Market Basket Analysis – The Apriori Algorithm. The Science of Data. Published October 27, 2019. Accessed March 17, 2022. https://scienceofdata.org/2019/10/27/market-basket-analysis-the-apriori-algorithm/

<p/>

Korstanje, J. (2021, September 24). The Apriori algorithm. Medium. https://towardsdatascience.com/the-apriori-algorithm-5da3db9aea95

<p/>

Market basket Analysis | Guide on Market Basket Analysis. (2021, October 2). Analytics Vidhya. https://www.analyticsvidhya.com/blog/2021/10/a-comprehensive-guide-on-market-basket-analysis/

<p/>

python - Apriori rule to pandas dataframe. (n.d.). Stack Overflow. Retrieved March 26, 2022, from https://stackoverflow.com/questions/70256325/apriori-rule-to-pandas-dataframe

<p/>

SHAH T. Market Basket Analysis. Nerd For Tech. Published July 4, 2021. Accessed March 16, 2022. https://medium.com/nerd-for-tech/market-basket-analysis-1c38613fdd6b

